### PR TITLE
Static Allocation

### PR DESCRIPTION
**Pull request by konbraphat51**
## What you changed (What reviewers have to see)
* Flow of making Static Allocation data #6
* Algorithm of Magnetic Allocation
* test codes of above
* Plenty of refacs

## Description
* Not implementing AnimatedAllocationCalculator.py and AllocationCalculator.py yet

* Haven't manually & visually checked the performance of the Magnetic Allocation, will check it and improve after this PR

## Check list
- [x] Code styles confirmed
- [x] Test confirmed
- [x]  Applied Vector class #33

### CODE REVIEW

File: AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationStrategies/RandomAllocation.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    :return: Left-top position of the word
+    :rtype: tuple[int, int]
+    """
+    # calculate the radius of the circle
+    image_diagnal = (image_width**2 + image_height**2) ** 0.5
+    text_diagnal = (word_size[0] ** 2 + word_size[1] ** 2) ** 0.5
+    radius = (image_diagnal + text_diagnal) / 2
+
+    # randomly choose where on the circle to put
+    angle = random() * 2 * pi
```
[the remark this answers is not in the document]
Comment (konbraphat51) [reply AnimatedWordCloud/Animator/AllocationCalculator/StaticAllocationStrategies/RandomAllocation.py:78]: Disabled

**Comment by konbraphat51** [comment]
The get_magnet_outer_frontier() function has a high cognitive complexity due to the nested loops and repeated code. Here's a plan to reduce its complexity:

Extract the repeated code into a separate function. This code is responsible for iterating over the x and y coordinates and checking if a point hits any rectangles. This function can take the range of x and y coordinates, the direction of y iteration, and the list of rectangles as parameters.

Use this new function in get_magnet_outer_frontier(). Replace the repeated code with calls to this function.

Here's the refactored code:

```python
def iterate_coordinates_and_check_hits(
    x_range: range,
    y_range: range,
    rects_outermost: Iterable[Rect]
) -> tuple[list[tuple[int, int]], set[Rect]]:
    new_hitted_rects = set()
    points = []
    for x in x_range:
        for y in y_range:
            flag_hitted, hitted_rect = is_point_hitting_rects((x, y), rects_outermost)
            if flag_hitted:
                new_hitted_rects.add(hitted_rect)
                points.append((x, y))
                break
    return points, new_hitted_rects

def get_magnet_outer_frontier(
    rects_outermost: Iterable[Rect],
    image_width: int,
    image_height: int,
    image_division: int,
) -> tuple[MagnetOuterFrontier, set[Rect]]:
    X_INTERVAL = image_width / image_division
    Y_INTERVAL = image_height / image_division

    magnet_outer_frontier = MagnetOuterFrontier()

    magnet_outer_frontier.from_up, new_hitted_rects_up = iterate_coordinates_and_check_hits(
        range(1, image_width, X_INTERVAL),
        range(0, image_height + 1, Y_INTERVAL),
        rects_outermost
    )

    magnet_outer_frontier.from_down, new_hitted_rects_down = iterate_coordinates_and_check_hits(
        range(1, image_width, X_INTERVAL),
        range(image_height, -1, -Y_INTERVAL),
        rects_outermost
    )

    new_hitted_rects = new_hitted_rects_up.union(new_hitted_rects_down)

    return magnet_outer_frontier, new_hitted_rects
```

This refactoring reduces the cognitive complexity by removing the nested loops and repeated code from get_magnet_outer_frontier(). The logic for iterating over the coordinates and checking for hits is now encapsulated in iterate_coordinates_and_check_hits().

**Review by SuperHotDogCat — COMMENTED** [review]
Default argument value should be needed for max_words, max_font_size, min_font_size,image_width,image_height,and font_path.
Default argument values should be set to the most high-level function.

File: AnimatedWordCloud/Utils/Vector.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+from __future__ import annotations
+
+
+class Vector:
+    """
+    Class for 2D vector calculation
+    """
+
+    def __init__(
+        self, x_or_tuple: float | tuple[float, float], y: float = 0
```
[the remark this answers is not in the document]
Comment (konbraphat51) [reply AnimatedWordCloud/Utils/Vector.py]: fixed

how is it?

File: AnimatedWordCloud/Utils/Vector.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+    def __init__(self, *args: float | tuple[float, float]) -> None:
+        """
+        Initialize Vector class.
+
+        :param args: x and y or tuple of x and y
+        input as Vector(x, y) or Vector((x, y))
+        """
+        if args[0].__class__ == tuple:
+            self.x = args[0][0]
```
Comment (SuperHotDogCat) [inline AnimatedWordCloud/Utils/Vector.py:26]: Ensure len(args[0]) == 2

Comment (SuperHotDogCat) [reply AnimatedWordCloud/Utils/Vector.py:26]: or Ensure len(args[0]) >= 2

Comment (SuperHotDogCat) [reply AnimatedWordCloud/Utils/Vector.py:26]: I don't know how best it is to fix it.
Maybe make sure that args[0][0] is a NUMBER and args[0][1] is a NUMBER

Comment (konbraphat51) [reply AnimatedWordCloud/Utils/Vector.py]: True,
I gave really better error handling

**Comment by konbraphat51** [comment]
OK

File: AnimatedWordCloud/Utils/Vector.py [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        Initialize Vector class.
+
+        :param args: x and y or tuple of x and y
+        input as Vector(x, y) or Vector((x, y))
+        """
+
+        # if the input is tuple...
+        if args[0].__class__ == tuple:
+            # ...ensuring (float, float)
+            if len(args[0]) != 2:
```
Comment (SuperHotDogCat) [inline AnimatedWordCloud/Utils/Vector.py:29]: OK

**Review by SuperHotDogCat — APPROVED** [review]
LGTM

**Pull request merged**